File: app/src/main/java/com/example/aqqhome/SignupActivity.java
package com.example.aqqhome;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.example.aqqhome.model.usermodel2;
import com.example.aqqhome.retrofit.ApiAQQHome;
import com.example.aqqhome.retrofit.RetrofitClient;
import com.google.firebase.auth.FirebaseAuth;

import java.util.Random;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class SignupActivity extends AppCompatActivity {
    private EditText emaillo, passwordlo, repasswordlo, namelo, phonelo;
    private Button signuplo;
    private FirebaseAuth mAuth;

    private Retrofit retrofit;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_signup);
        emaillo = findViewById(R.id.emaillo);
        passwordlo = findViewById(R.id.passwordlo);
        repasswordlo = findViewById(R.id.repasswordlo);
        namelo = findViewById(R.id.namelo);
        phonelo = findViewById(R.id.phonelo);
        signuplo = findViewById(R.id.signuplo);

        mAuth = FirebaseAuth.getInstance();
        retrofit = RetrofitClient.getRetrofitInstance();

        signuplo.setOnClickListener(v -> {
            String email = emaillo.getText().toString();
            String pwd = passwordlo.getText().toString();
            String repwd = repasswordlo.getText().toString();
            String name = namelo.getText().toString();
            String phone = phonelo.getText().toString();

            if (email.isEmpty()) {
                emaillo.setError("Vui lòng nhập email");
                emaillo.requestFocus();
            } else if (pwd.isEmpty()) {
                passwordlo.setError("Vui lòng nhập mật khẩu");
                passwordlo.requestFocus();
            } else if (repwd.isEmpty()) {
                repasswordlo.setError("Vui lòng nhập lại mật khẩu");
                repasswordlo.requestFocus();
            } else if (name.isEmpty()) {
                namelo.setError("Vui lòng nhập tên");
                namelo.requestFocus();
            } else if (phone.isEmpty()) {
                phonelo.setError("Vui lòng nhập số điện thoại");
                phonelo.requestFocus();
            } else if (pwd.length() < 6) {
                passwordlo.setError("Mật khẩu không được dưới 6 kí tự");
                passwordlo.requestFocus();
            } else if (repwd.length() < 6) {
                repasswordlo.setError("Mật khẩu không được dưới 6 kí tự");
                repasswordlo.requestFocus();
            } else if (pwd.equals(repwd)) {
                Random random = new Random();
                String uid = String.valueOf(random.nextInt(20));
                PostMysql(email, pwd, name, phone, uid);
                Intent intent = new Intent(SignupActivity.this, LoginActivity.class);
                startActivity(intent);

            } else {
                repasswordlo.setError("Mật khẩu không hợp lệ");
                repasswordlo.requestFocus();
            }
        });
    }

    private void PostMysql(String email, String pwd, String name, String phone, String uid) {
        ApiAQQHome ApiAQQHome = retrofit.create(ApiAQQHome.class);
        Call<usermodel2> call = ApiAQQHome.dangki(email, pwd, name, phone, uid);
        call.enqueue(new Callback<usermodel2>() {
            @Override
            public void onResponse(Call<usermodel2> call, Response<usermodel2> response) {
                if (response.isSuccessful()) {
                    Toast.makeText(SignupActivity.this, "Đăng ký thành công", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(SignupActivity.this, "Máy chủ không phản hồi xin vui lòng quay lại sau", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<usermodel2> call, Throwable t) {
                Toast.makeText(SignupActivity.this, "Vui lòng bật lại mạng", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/example/aqqhome/SignupActivity.kt
package com.example.aqqhome

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.aqqhome.model.usermodel2
import com.example.aqqhome.retrofit.ApiAQQHome
import com.example.aqqhome.retrofit.RetrofitClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class SignupActivity : AppCompatActivity() {

    private lateinit var emailInput: EditText
    private lateinit var passwordInput: EditText
    private lateinit var repasswordInput: EditText
    private lateinit var nameInput: EditText
    private lateinit var phoneInput: EditText
    private lateinit var signUpButton: Button
    private val retrofit by lazy { RetrofitClient.getRetrofitInstance() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_signup)

        initializeViews()
        setupSignUpButton()
    }

    private fun initializeViews() {
        emailInput = findViewById(R.id.emaillo)
        passwordInput = findViewById(R.id.passwordlo)
        repasswordInput = findViewById(R.id.repasswordlo)
        nameInput = findViewById(R.id.namelo)
        phoneInput = findViewById(R.id.phonelo)
        signUpButton = findViewById(R.id.signuplo)
    }

    private fun setupSignUpButton() {
        signUpButton.setOnClickListener {
            val email = emailInput.text.toString()
            val password = passwordInput.text.toString()
            val repassword = repasswordInput.text.toString()
            val name = nameInput.text.toString()
            val phone = phoneInput.text.toString()

            when {
                email.isBlank() -> emailInput.setErrorAndFocus("Vui lòng nhập email")
                password.isBlank() -> passwordInput.setErrorAndFocus("Vui lòng nhập mật khẩu")
                repassword.isBlank() -> repasswordInput.setErrorAndFocus("Vui lòng nhập lại mật khẩu")
                name.isBlank() -> nameInput.setErrorAndFocus("Vui lòng nhập tên")
                phone.isBlank() -> phoneInput.setErrorAndFocus("Vui lòng nhập số điện thoại")
                password.length < 6 -> passwordInput.setErrorAndFocus("Mật khẩu không được dưới 6 kí tự")
                repassword.length < 6 -> repasswordInput.setErrorAndFocus("Mật khẩu không được dưới 6 kí tự")
                password != repassword -> repasswordInput.setErrorAndFocus("Mật khẩu không hợp lệ")
                else -> {
                    val uid = UUID.randomUUID().toString()
                    postToMysql(email, password, name, phone, uid)
                    startActivity(Intent(this, LoginActivity::class.java))
                }
            }
        }
    }

    private fun EditText.setErrorAndFocus(errorMsg: String) {
        this.error = errorMsg
        this.requestFocus()
    }

    private fun postToMysql(email: String, password: String, name: String, phone: String, uid: String) {
        val apiAQQHome = retrofit.create(ApiAQQHome::class.java)
        val call = apiAQQHome.dangki(email, password, name, phone, uid)

        call.enqueue(object : Callback<usermodel2> {
            override fun onResponse(call: Call<usermodel2>, response: Response<usermodel2>) {
                val message = if (response.isSuccessful) {
                    "Đăng ký thành công"
                } else {
                    "Máy chủ không phản hồi xin vui lòng quay lại sau"
                }
                Toast.makeText(this@SignupActivity, message, Toast.LENGTH_SHORT).show()
            }

            override fun onFailure(call: Call<usermodel2>, t: Throwable) {
                Toast.makeText(this@SignupActivity, "Vui lòng bật lại mạng", Toast.LENGTH_SHORT).show()
            }
        })
    }
}
